src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -389,7 +398,8 @@ private void signalIfAllRegisteredInQueue() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size()) {
+        if (registeredElements.size() == liveQueue.size() ||
+                poisonPill != null) {
```

Comment:
Since the `allRegisteredInQueue` Condition variable is basically only used in the context of signalling a waiting locker that it can proceed, maybe renaming this to something like `lockAvailable` or something like that would be better, or at least shorter anyway.